### ISSUE TITLE
Button presses getting stuck when using keyboard

### ISSUE THREAD

**Issue by ShadowStkr** [visualboyadvance-m/visualboyadvance-m#142]
### Expected behavior

Once a key on the keyboard which is configured to a certain button is released, it shouldn't be registered as released and not being pressed/held anymore.

### Actual behavior

Once releasing a certain key (say, left directional, which is configured to the left button), at most times the corresponding button is also registered as not being pressed anymore. However, at certain times (which seem to be random), even after being released, the button corresponding to the key which was pressed still registers as if it was being pressed/hold. If the key is pressed and released again, its release is registered and everything comes back to normal.

I've tried reproducing the problem with a Dualshock 3 (aka PS3 controller) but it seems to not happen with it. Also, something curious is that if I press and release the button on the DS3 which corresponds in the input options to the same button which got "stuck" when using the keyboard, it doesn't get registered as released, as opposed to when doing the same in the keyboard itself.

### Steps to reproduce the behavior

Either press a key a certain number of times with some pauses between the presses to check whether its input got "stuck" or alternate between buttons until one of them gets stuck.  It seems to be easier to reproduce by pressing directional keys randomly on a menu screen with a lot of options which can be switched horizontally and vertically (such as a naming screen in a game, for example). Turbo mode may also help.

### ROM details

I experienced this with Harvest Moon: Friends of Mineral Town (US version). I was able to reproduce the problem in both the initial naming screen and while controlling the character at random. I also tried it with the english version of Golden Sun, in the initial naming screen, and was able to reproduce the problem, so I assume it's not a problem limited to a certain ROM.

### Option details

Video driver (simple, opengl or cairo): OpenGL

Sound driver (directsound, xaudio, openal, SDL): Problem happens with both OpenAL/SDL

### Build details

Commit 4d7e102e2650d9d66ac0ef49fac3869dd29df8da, built and tested on Arch Linux. Using xorg.

**Comment by rkitover** [comment]
Possibly will need to inspect physical keyboard on key events, maybe even switch to using SDL for keyboard game keys.

SDL is already being used for joystick events.

**Comment by Yoshimaster96** [comment]
This happens to me too. I'm on Linux Mint Cinnamon 64-bit, using VBA-M 2.0.0 beta 3.
The game doesn't seem to matter, and it happens on GB, GBC, and GBA games.
Video: (Simple/OpenGL): OpenGL
Audio: (SDL/OpenAL): OpenAL

**Comment by rkitover** [comment]
@Yoshimaster96 please build from source or use the ubuntu nightly PPA.

**Comment by Yoshimaster96** [comment]
@rkitover The nightly build still has the bug.
Building gives me this error when running:
```Fatal Error: Mismatch between the program and library build versions detected.
The library used 3.0 (wchar_t,compiler with C++ ABI 1002,wx containers,compatible with 2.8),
and your program used 3.0 (wchar_t,compiler with C++ ABI 1009,wx containers,compatible with 2.8).
Aborted
```

**Comment by ZachBacon** [comment]
Sounds like your distro is using a version of wxwidgets that differs from the one the PPA compiles against.

**Comment by ArtiiP** [comment]
@Yoshimaster96 try add -fabi-version=2 to GCCFLAGS(or something ;p ) and full clean&build
https://gcc.gnu.org/onlinedocs/gcc-7.2.0/gcc/C_002b_002b-Dialect-Options.html

it work's for me, but i have 1010 vs 1011 (and in another project).

**Comment by rkitover** [comment]
@ArtiiP thanks for the info! I added the flag to the cmake code for gcc.

**Comment by rkitover** [comment]
@Yoshimaster96 try backing up and deleting your config file:

```shell
mv ~/.vbam/vbam.conf ~/.vbam/vbam.conf.old
```

let me know if that helps. If it does, please post your old config file here.

**Comment by ArtiiP** [comment]
> @ArtiiP thanks for the info! I added the flag to the cmake code for gcc.

but you know, this broke for people with wx compiled with other abi version ;/

it's not easy fix,  maybe expose opiton (and info) for cmake 
(ofc better fix is autodetection, but i don't know how, and google don't help either)

**Comment by Yoshimaster96** [comment]
Yup, bug still occurs when building from source.

**Comment by rkitover** [comment]
@Yoshimaster96 can you try removing your config file? See above.

**Comment by rkitover** [comment]
@ArtiiP I will write a compile/link/run test to determine the correct ABI version later on today and will let you know, thanks!

**Comment by rkitover** [comment]
@ArtiiP got it working, see 2ad2046. Yes cmake is an ugly language 😃

**Comment by Yoshimaster96** [comment]
Please consider fixing the button issue however. I am still experiencing this bug.

**Comment by Yoshimaster96** [comment]
@rkitover What's with the delay?

**Comment by ZachBacon** [comment]
The delay is that some of us have a life. I work full time at a warehouse and rkitover has things and other projects that take his attention. We do this on our spare time. Please continue being patient, we appreciate it :)

**Comment by Yoshimaster96** [comment]
@ZachBacon Ah okay, sorry about that. I just thought it was forgotten about entirely.

**Comment by rkitover** [comment]
@Yoshimaster96 we'll need to be able to reproduce the bug in order to do something about it, right now I don't know how to trigger it.

**Comment by Yoshimaster96** [comment]
@rkitover So it's never happened to you?

**Comment by ale64bit** [comment]
I experience this to the point of rendering games unplayable.

For example, how to reproduce it playing Metroid: Zero Mission: just move towards a wall in your left, and then hold left for about 5 seconds and release. Under my setup, many times the key release doesn't seem to be registered and the character continues to run left until the left key is pressed again.

I'm on Ubuntu 18.04, Thinkpad 470, using keyboard, installed from PPA (v2.0.2).
Thanks!

**Comment by Arararararagi** [comment]
This happens to me while playing Mario Kart. Pretty freaking annoying when I don't even know which key got stuck anymore.

Machine: Ideapad 300

OS: Linux 4.14.31-1-MANJARO #1 SMP PREEMPT Wed Mar 28 21:42:49 UTC 2018 x86_64 GNU/Linux

VisualBoyAdvance-M (vbam-wx)  2.0.1-1, installed from package manager (Community repository).

**Comment by manugarri** [comment]
Same issue here, ubuntu 18.04 with corsair keyboard.

**Comment by johtull** [comment]
I am also getting this issue on Ubuntu 18.04.

A workaround for me has been changing the Output module from OpenGL to Simple.

I'm on version 2.1.0-b4dd06a1 compiled from source.

**Comment by DataFlame** [comment]
I'm having this problem, I tried the workaround that johtull mentioned but that still didn't work.

I'm on version 2.1.0 and I'm using Linux Mint 19.

**Comment by MadaraUchiha** [comment]
This reproduced on my end. VisualBoyAdvance-M 2.1.0-5f38c0da on Ubuntu 18.10.

**Comment by apocalyptech** [comment]
I had this happen to me on Metroid Zero Mission, VBA-M v2.1.0...  In my case it looks like it might have been related to having my "fire" button mapped to Ctrl, though.  When I changed that bind to a "real" key, I haven't had the sticky-direction problem for a good ten minutes now, when I used to be having it at least once a minute.

Obviously CTRL can end up with some more complicated keyboard signals and the like, so it at least makes sense that that could be related.  Anyway, I'll edit this if moving away from CTRL ends up *not* actually fixing it after all.  (This obviously wouldn't explain some of the other reproduction cases above, where it seemed that only direction keys were involved...)

**Comment by retro-wertz** [comment]
> I had this happen to me on Metroid Zero Mission, VBA-M v2.1.0... In my case it looks like it might have been related to having my "fire" button mapped to Ctrl, though. When I changed that bind to a "real" key, I haven't had the sticky-direction problem for a good ten minutes now, when I used to be having it at least once a minute.
> 
> Obviously CTRL can end up with some more complicated keyboard signals and the like, so it at least makes sense that that could be related. Anyway, I'll edit this if moving away from CTRL ends up _not_ actually fixing it after all. (This obviously wouldn't explain some of the other reproduction cases above, where it seemed that only direction keys were involved...)
> 
> Edit: I actually seemed to have some similar issues when attempting to use SHIFT as a keybind. That would similarly screw things up on occasion. I've not had the problem reproduce after about 30min of playing, though, when I've got only "regular" keys bound, so that's still looking good.

if you are able to compile from source, please try latest commit. if not, ill see if they can provide nightly/sub release with the fix to inputs.

**Comment by apocalyptech** [comment]
> if you are able to compile from source, please try latest commit. if not, ill see if they can provide nightly/sub release with the fix to inputs.

Sure thing, compiling isn't a problem.  On 2.1.0-6330555c, I definitely still have that problem when trying to use CTRL as one of my keybinds.  With CTRL bound to the fire button (B) on Metroid Zero Mission, for instance, just hitting that at the same time as moving left or right a few times will generally trigger it reliably within a few seconds.  Reverting to using a "real" key instead of CTRL clears that up for me.

Let me know if I can try anything else, or provide logs or whatever!

**Comment by nemec** [comment]
I have the exact same issue. I just tried the latest master https://github.com/visualboyadvance-m/visualboyadvance-m/commit/6330555c3bc5d54ffd4252911e939691f136a68a and it did not fix the issue, nor did clearing my config file (the issue was present even without touching the default button configuration).

I can reliably reproduce it in Fire Emblem 7 (the English version). Simply go onto a map and hold one of the directional arrow keys for 3-4 seconds. Often once you release the key your cursor will continue moving in the same direction until it hits the end of the map, and depending on directional priority it can prevent you from even moving the cursor in the opposite direction.

I tried enabling verbose logging, but the only output was one "Battery read" message and one "Battery write" message.

**Comment by NoSuck** [comment]
I'm glad I'm not the only one affected by this issue.  I'll try building from the latest `master` and see if the stuck keyboard key issue disappears.  In the meantime, I'm using 2.1.0 on:

```
Linux 4.20.3-arch1-1-ARCH #1 SMP PREEMPT Wed Jan 16 22:38:58 UTC 2019 x86_64 GNU/Linux
```

**Comment by Xetr0n** [comment]
I know how to reproduce this bug in The Legend of Zelda: The Minish Cap.
I binded moving to W,A,S,D and A to Shift and B to Space. When I now press a button to walk and hold it and then press Shift (A), hold it and release the walking button, then it gets stuck. I can also reproduce this using CTR which is not even binded to anything for me.

I'm using VisualBoyAdvance-M 2.1.0 on Windows 10 64bit

**Comment by rkitover** [comment]
@Xetr0n can you please try the latest nightly and see if the problem persists?

**Comment by retro-wertz** [comment]
Nightly release:
https://github.com/visualboyadvance-m/visualboyadvance-m/issues/354

**Comment by Xetr0n** [comment]
@rkitover yes tried it there (10.02.2019) got a 100% reproduction rate with all modifier keys (LSHIFT, RSHIFT, RCTRL, ALT, [on german keyboard "Alt Gr", a combination of ALT and CTRL]). Just hold a key and then press and hold one modifier key and release the "normal" key.

**Comment by rkitover** [comment]
@Xetr0n thanks!

**Comment by NoSuck** [comment]
For what it's worth, I have been unable to reproduce this issue with [the libretro version of vba-m](https://github.com/libretro/vba-next).

**Comment by retro-wertz** [comment]
> For what it's worth, I have been unable to reproduce this issue with [the libretro version of vba-m](https://github.com/libretro/vba-next).

libretro is a different frontend/api.. the issue here is regarding the standalone frontend...

**Comment by NoSuck** [comment]
... none of which matters to someone wanting to play a game.

**Comment by retro-wertz** [comment]
whats your point then...

**Comment by Kiwii** [comment]
I can reproduce the issue on Linux with version 2.1.0 from Arch's repos as well as v2.1.1-0-gb5741ee4 compiled from git, using Xetr0n's instructions: 

1. hold a modifier (e.g. CTRL)
2. hold another button (e.g. left arrow)
3. release the modifier
4. release the other button

Expected behavior: Character stops walking.
Actual behavior: Character continues walking.

I can not reproduce the issue using a controller.

**Comment by NoSuck** [comment]
> whats your point then...

To let users who come across this unresolved issue from 2017 know that an alternative exists.

**Comment by BurningTurtle** [comment]
Any news on this?

**Comment by denisfa** [comment]
@BurningTurtle We can confirm this still happens on keyboard. Not so sure about the fix, though. I will give it a try after my ongoing work.

**Comment by rkitover** [comment]
Please try latest nightly.

**Comment by huupoke12** [comment]
Commit 3d57750c16e28d834e6c210cc5d93d14a780e1e0 has fixed the issue with the modifier key. But the original issue is not fixed. The buttons sometimes get stuck (It can stuck even with only default directional keys) so this is not the same bug.

**Comment by denisfa** [comment]
@huupoke12 Thanks for the feedback! Are you able to reproduce this bug? I can't reproduce this at the moment. I tried the game `Harvest Moon - Friends Of Mineral Town`, but it was not successful.

If so, would you mind providing the name of the game, OS you are running and your `vbam.ini` file so I can investigate what's going on?

The controller code was refactored not so long ago, so I believe this should be a keyboard-only bug.

**Comment by huupoke12** [comment]
The games I've tried are: `Mario Kart - Super Circuit, SpongeBob SquarePants - Battle for Bikini Bottom,
Pokemon Leaf Green`
I've also tried `Harvest Moon - Friends Of Mineral Town`, on the player name selection screen, the bug still happens. You can just hold right key for about 2 seconds, about 10% it will be stuck.
I'm using Arch Linux (5.1.4-arch1-1-ARCH).
[vbam.zip](https://github.com/visualboyadvance-m/visualboyadvance-m/files/3219916/vbam.zip)

**Comment by nemec** [comment]
I just tried the latest master and still have the same issue. Using the original (U.S.) Fire Emblem game (my rom is named `1235 - Fire Emblem (U)(Venom).gba`), place the attached `.sav` and `.sgm` files into the same dir and load the game, then load the save state. Your cursor will show up on the far right of the screen. Simply hold the left arrow key until the cursor _almost_ hits the end of the screen, then let go. Repeat left and right as needed until the cursor "sticks" in one direction. This works for me ~20% of the time in any one direction (100% after going left to right ~5 times)

Ubuntu 18.04.2 LTS

[FE-saves.zip](https://github.com/visualboyadvance-m/visualboyadvance-m/files/3219926/FE-saves.zip)

[vbam.ini](https://gist.github.com/nemec/6258fbddcca96c2641e8ca88f269faf7)

**Comment by rkitover** [comment]
Does this only happen if you load a save state?

**Comment by huupoke12** [comment]
Maybe not. In my case, I played `Harvest Moon - Friends Of Mineral Town` the first time to check for the bug so there won't be any save file before.

**Comment by nemec** [comment]
Not for me, it happens regardless of whether I loaded a save state or not. I just uploaded the save to provide a quick starting point in case you wanted to try to repro the issue. I'm happy to send the rom, too, if you message me on the email in my profile. There is nothing special about it, though. The saves should work with any FE7 from a GBA rom collection.

**Comment by denisfa** [comment]
@huupoke12 @nemec Thank you for this feedback! I am afraid I have had a lot of luck on my games, since I failed both `Harvest Moon - Friends Of Mineral Town` and `Fire Emblem`, even with the supplied save state... There is nothing special about your config files, so this already removes some possibilities.

>Your cursor will show up on the far right of the screen. Simply hold the left arrow key until the cursor almost hits the end of the screen, then let go. Repeat left and right as needed until the cursor "sticks" in one direction. This works for me ~20% of the time in any one direction (100% after going left to right ~5 times)

This is pretty straightforward, the 20% of any against the 100% from left to right is way too weird.

I noticed that the LEFT_KEY takes over the RIGHT_KEY when I accidentally pressed them both at the same time. Just putting this here so I can remember later.

@rkitover while trying to check this, I (think I) discovered an issue with the pause from the frontend. Could you see https://github.com/visualboyadvance-m/visualboyadvance-m/commit/e4cb6787d6762ffb5a25973ccaa91e59107ca810 and confirm If this is an issue indeed?

**Comment by nemec** [comment]
> I noticed that the LEFT_KEY takes over the RIGHT_KEY when I accidentally pressed them both at the same time. Just putting this here so I can remember later.

I noticed that, too. In Fire Emblem this means that a stuck LEFT_KEY forces the cursor to the left side of the screen - you can move up and down, but you are unable to move right until you hit LEFT_KEY again to unstick it. And if RIGHT_KEY is stuck, your cursor will "roll" toward the right side of the screen, but pressing LEFT_KEY will override the stuck key and move your cursor left. Releasing LEFT_KEY will then cause the cursor to roll back to the right, since RIGHT_KEY is still stuck.

I [recorded](https://vimeo.com/338502271) a screencast of both examples. At 00:07, RIGHT_KEY is already stuck. You can see me hit LEFT_KEY a few times (and only left), then let the keyboard sit. The cursor moves right without any additional input from me. At 00:15 I end up getting LEFT_KEY stuck. You can see me push RIGHT_KEY a few times at 00:18 but my cursor stays on the left wall without moving.

**Comment by denisfa** [comment]
What Desktop Environment are you using? We are trying to check something if there might be something related to this bug.

**Comment by BurningTurtle** [comment]
Yes, it still happens.

I'm using Ubuntu 18.04 with Pokemon Emerald.

**Comment by denisfa** [comment]
@huupoke12 , @nemec, @BurningTurtle Would you mind testing a possible solution for this? It should be like this:

```
cd ~ && mkdir src && cd src
git clone -b fix-142 https://github.com/visualboyadvance-m/visualboyadvance-m.git
cd visualboyadvance-m
./installdeps

# ./installdeps will give you build instructions, which will be similar to:

mkdir build && cd build
cmake ..
make -j`nproc`
./visualboyadvance-m
```

This is the branch with the possible fix. I managed to set up a VM and finally be able to run into this issue consistently. Using this branch fixed the issue for me.

**Comment by nemec** [comment]
That's perfect, @denisfa! I built it and moved the cursor back and forth for ~3 minutes and did not see the issue once. Thank you for tracking this down!

**Comment by huupoke12** [comment]
I have tested and didn't run to any issue with it with the default config. It only happens when I hold a key, then click outside of the windows (still hold the key) and release it when "Pause when inactive" is unticked.

**Comment by denisfa** [comment]
@nemec Good to know! I believe this should be ready to land soon! Thanks for the feedback, it really helped track this down.

> I have tested and didn't run to any issue with the default config. It only happens when I hold a key, then click outside of the windows (still hold the key) and release it when "Pause when inactive" is unticked.

@huupoke12 I found out about this issue while trying to run into the original bug. It should also be fixed. If you don't mind, you can test the given branch solution for this one too.

**Comment by huupoke12** [comment]
I've tested commit 42d3577647a960ef3dd339b719ec34790ab29246, the bug has been fixed.

**Comment by denisfa** [comment]
@huupoke12 Thank you for the testing! I will consider it fixed. It will be available for our next release `2.1.4`.
